Add support for DSim simulator

teobiton opened the pull request:
Support DSim in the Makefile flow, with split compile and run phases and with wave generation support.

Closes #3990

ktbarrett reviewed (commented):
Off the bat I think we are going to need a Python runner implementation, this should be added to `noxfile.py` so it can be supported for regression and packaging, and it would be nice to see this running in CI, if possible.

ktbarrett commented:
> I don't know about CI since a license is required to run the simulations.

Oh well... perhaps we can talk them into giving us one, but that doesn't need to block this. Just the runner then.

ktbarrett reviewed (commented):
Looks good overall. Can you post some screenshots of it working in lieu of a regression test?

cmarqu commented on docs/source/newsfragments/3990.feature.rst:
```diff
@@ -0,0 +1 @@
+Altair DSim simulator is now supported by cocotb.
```
```suggestion
The Siemens DSim simulator is now supported by cocotb.
```
(not sure if the sale is completed yet)

teobiton replied:
I guess it is: https://press.siemens.com/global/en/pressrelease/siemens-acquires-altair-create-most-complete-ai-powered-portfolio-industrial-software

I changed Altair to Siemens but the links to install Dsim are still on Altair Marketplace.

imphil commented:
Thanks for your PR, @teobiton! 

* Can you also add the required steps to run the simulation to https://github.com/cocotb/cocotb/blob/master/docs/source/custom_flows.rst? That page also serves as good reference for us.
* Please add dsim also to https://github.com/cocotb/cocotb/blob/master/noxfile.py#L43 and run the cocotb tests against dsim. How well does it work? (You don't need to fix all issues, but it would be good to get a first feeling).

Andrea-Spitale commented:
Hi,
Just wanted to have you know that dsim was bought by Altair, and is now available (maybe in a feature restricted way ) at [https://learn.altair.com/course/view.php?id=810&_gl=1*1rpezuc*_gcl_au*MjAzMDIwNjg5Ny4xNzQzNDI0MjQ5](this link). Might be useful to have cocotb users aware.

teobiton commented:
Hi all, thanks for your feedback on the PR!

Here are a few screenshots using Dsim on the simple dff example.

@imphil I added dsim in noxfile and ran the tests. I identified a few tests where it fails but the main issue is that if the simulator meets an internal error and stops abnormally, there is like a 5 minutes delay before the license is un-issued and available again, so all tests remaining in the regression are not actually run. I started a local report of all issues and I'm adding skips to tests that cause internal errors.

@Andrea-Spitale thank you for the link, I added it the docs.

teobiton commented on src/cocotb_tools/runner.py:
From https://help.metrics.ca/support/solutions/articles/154000141196, I think you're right.

ktbarrett commented on src/cocotb_tools/makefiles/simulators/Makefile.dsim:
```diff
@@ -0,0 +1,42 @@
+	$(SIM_CMD_PREFIX) $(CMD) $(RUN_ARGS) $(TOPMODULE_ARG) $(EXTRA_ARGS) $(VERILOG_SOURCES)
+
+	$(call check_results)
+
+clean::
+	$(RM) -r $(SIM_BUILD)
```
This is covered in Makefile.inc.
```suggestion
```

ktbarrett commented on src/cocotb_tools/runner.py:
Just checking this is correct? You would be the one that knows. Do the include paths need to be formatted as TCL strings?

teobiton replied:
Nop, this was a mistake.

ktbarrett commented on src/cocotb_tools/runner.py:
This should probably be `NotImplementedError`. If the other sims use `ValueError` you can just ignore this suggestion and we'll clean them all up at the same time later.

teobiton replied:
Hm, some simulators have `RuntimeError` and others have `ValueError`.

ktbarrett replied:
Oh well.

ktbarrett reviewed (approved):
One nitpick and some questions.

ktbarrett commented:
I'll give it a day for other maintainers to chime in then merge.

ktbarrett commented:
Thanks again @teobiton!
